### PR TITLE
PORTALS-2157: Refactor TotalQueryResults and SelectionCriteriaPill

### PR DESCRIPTION
**Pull request by nickgros**
- Refactor TotalQueryResults and SelectionCriteriaPill
- Query modification (e.g. dropping facet selection) is managed by useImmutableTableQuery
- Displaying a column name is managed by QueryVisualizationWrapper
- Remove prop `idForTooltip` from ElementWithTooltip (no longer needed)

### CODE REVIEW

File: src/lib/containers/QueryVisualizationWrapper.tsx [hunk 1]
```diff
@@ -28,6 +30,7 @@ export type QueryVisualizationContextType = {
   showLastUpdatedOn?: boolean
   /** Given a column name, return the display name for the column */
   getColumnDisplayName: (columnName?: string) => string | undefined
+  getDisplayValue: (value: string, columnType: ColumnType) => string
```
Comment (nickgros) [inline src/lib/containers/QueryVisualizationWrapper.tsx:33]: Shared way to transform cell data -> displayed value

File: src/lib/containers/QueryVisualizationWrapper.tsx [hunk 2]
```diff
@@ -12,6 +12,8 @@ import { NoContentPlaceholderType } from './table/NoContentPlaceholderType'
 import SearchResultsNotFound from './table/SearchResultsNotFound'
 import ThisTableIsEmpty from './table/TableIsEmpty'
 import { unCamelCase } from '../utils/functions/unCamelCase'
+import { ColumnType } from '../utils/synapseTypes'
+import { getDisplayValue } from '../utils/functions/getDataFromFromStorage'
```
Comment (nickgros) [inline src/lib/containers/QueryVisualizationWrapper.tsx:16]: While this is imported for now, we can change it (e.g. to use react-query) later.

File: src/lib/containers/TotalQueryResults.tsx [hunk 3]
```diff
@@ -55,300 +18,19 @@ export type TotalQueryResultsProps = {
...
 
   const { topLevelControlsState, unitDescription } =
     useQueryVisualizationContext()
 
   const total = data?.queryCount
-  const [facetsWithSelection, setFacetsWithSelection] = useState<
-    FacetWithSelection[]
-  >([])
-
-  const applyChanges = (facets: FacetColumnRequest[]) => {
```
Comment (nickgros) [inline src/lib/containers/TotalQueryResults.tsx:84]: Applying changes to the query is now handled in `useImmutableTableQuery`

File: src/lib/containers/TotalQueryResults.tsx [hunk 4]
```diff
@@ -55,300 +18,19 @@ export type TotalQueryResultsProps = {
...
-
-  const getDisplayValueUserIdColumn = (
-    userProfiles: UserProfile[],
-    facetValue: string,
-  ): string => {
-    const userProfile = userProfiles.find(item => item.ownerId === facetValue)
-    return userProfile?.userName || facetValue
-  }
 
-  const transformEnumFacetsForSelectionDisplay = useCallback(
```
Comment (nickgros) [inline src/lib/containers/TotalQueryResults.tsx:140]: Now handled by QueryVisualizationWrapper.getDisplayValue

File: src/lib/containers/TotalQueryResults.tsx [hunk 5]
```diff
@@ -55,300 +18,19 @@ export type TotalQueryResultsProps = {
...
-    executeQueryRequest(cloneLastQueryRequest)
-  }
-
-  const clearAll = () => {
-    const initQueryRequest = cloneDeep(getInitQueryRequest())
-    initQueryRequest.query.additionalFilters = []
-    executeQueryRequest(initQueryRequest)
-  }
-
-  const searchSelectionCriteriaPillProps:
```
Comment (nickgros) [inline src/lib/containers/TotalQueryResults.tsx:305]: Transforming facets into pill props was cleaned up and moved to SelectionCriteriaPills.tsx

File: src/lib/containers/useImmutableTableQuery.ts [hunk 6]
```diff
@@ -88,74 +110,189 @@ export default function useImmutableTableQuery(
...
 
   /**
-   * Pass down a deep clone (so no side affects on the child's part) of the
+   * Pass down a deep clone (so no side effects on the child's part) of the
    * first query request made
    *
    * @returns
    * @memberof QueryWrapper
    */
-  function getInitQueryRequest(): QueryBundleRequest {
```
Comment (nickgros) [inline src/lib/containers/useImmutableTableQuery.ts:97]: useCallback-ified

File: src/lib/containers/useImmutableTableQuery.ts [hunk 7]
```diff
@@ -88,74 +110,189 @@ export default function useImmutableTableQuery(
...
-  }
+  }, [initQueryRequest])
 
   /**
-   * Execute the given query request, updating all of the data in the QueryContext to match the new query
+   * Execute the given query request, updating all the data in the QueryContext to match the new query
    * @param {*} queryRequest Query request as specified by
    *                         https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/table/Query.html
    */
-  function setQuery(
```
Comment (nickgros) [inline src/lib/containers/useImmutableTableQuery.ts:106]: useCallback-ified

File: src/lib/containers/useImmutableTableQuery.ts [hunk 8]
```diff
@@ -88,74 +110,189 @@ export default function useImmutableTableQuery(
...
+  const { entityId, versionNumber } = useMemo(
+    () => parseEntityIdAndVersionFromSqlStatement(lastQueryRequest.query.sql)!,
+    [lastQueryRequest.query.sql],
+  )
 
   const pageSize = lastQueryRequest.query.limit ?? DEFAULT_PAGE_SIZE
   const currentPage = Math.ceil(
     ((lastQueryRequest.query.offset ?? 0) + Number(pageSize)) / pageSize,
   )
-  function setPageSize(pageSize: number) {
```
Comment (nickgros) [inline src/lib/containers/useImmutableTableQuery.ts:146]: useCallback-ified

File: src/lib/containers/useImmutableTableQuery.ts [hunk 9]
```diff
@@ -88,74 +110,189 @@ export default function useImmutableTableQuery(
...
-    setQuery(currentQuery => {
-      currentQuery.query.limit = pageSize
-      return currentQuery
-    })
-  }
 
-  function goToPage(pageNumber: number) {
```
Comment (nickgros) [inline src/lib/containers/useImmutableTableQuery.ts:153]: useCallback-ified

File: src/lib/containers/useImmutableTableQuery.ts [hunk 10]
```diff
@@ -88,74 +110,189 @@ export default function useImmutableTableQuery(
...
+        currentQuery.query.offset = (pageNumber - 1) * pageSize
+        return currentQuery
+      })
+    },
+    [pageSize, setQuery],
+  )
+
+  const resetQuery = useCallback(() => {
+    setQuery(initQueryRequest)
+  }, [initQueryRequest, setQuery])
```
Comment (nickgros) [inline src/lib/containers/useImmutableTableQuery.ts:195]: New

File: src/lib/containers/widgets/facet-nav/SelectionCriteriaPill.stories.tsx [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+export default {
+  title: 'Explore/Tokens/SelectionCriteriaPill',
+  component: SelectionCriteriaPill,
+  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
+  argTypes: {},
+} as ComponentMeta<typeof SelectionCriteriaPill>
+
+// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
+const Template: ComponentStory<typeof SelectionCriteriaPill> = args => (
+  <SelectionCriteriaPill {...args} />
```
Comment (nickgros) [inline src/lib/containers/widgets/facet-nav/SelectionCriteriaPill.stories.tsx:15]: Adding a story for this (now-stateless) component.

File: src/lib/containers/widgets/facet-nav/SelectionCriteriaPill.tsx [hunk 12]
```diff
@@ -1,94 +1,28 @@
...
-      filter: {
-        columnName?: string
-        value: string
-      }
-    }
-) & {
-  onRemove: () => void
-}
-
-const SelectionCriteriaPill: FunctionComponent<
```
Comment (nickgros) [inline src/lib/containers/widgets/facet-nav/SelectionCriteriaPill.tsx:31]: This component is now much more simple

File: src/lib/containers/widgets/facet-nav/SelectionCriteriaPills.tsx [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+  FacetColumnRequest,
+  isFacetColumnRangeRequest,
+  isFacetColumnValuesRequest,
+} from '../../../utils/synapseTypes'
+import {
+  QueryVisualizationContextType,
+  useQueryVisualizationContext,
+} from '../../QueryVisualizationWrapper'
+
+function getPillPropsFromColumnQueryFilter(
```
Comment (nickgros) [inline src/lib/containers/widgets/facet-nav/SelectionCriteriaPills.tsx:25]: An individual QueryFilter/FacetColumnRequest may have multiple values, which means multiple pills should be shown. These functions control that. Hopefully much more clean than what we had in `TotalQueryResults` before this PR.

File: src/lib/utils/functions/getDataFromFromStorage.ts [hunk 14]
```diff
@@ -1,7 +1,12 @@
...
+import {
+  ColumnType,
+  EntityHeader,
+  Evaluation,
+  UserProfile,
+} from '../synapseTypes'
 import { SynapseConstants } from '..'
 
-export const getStoredEntityHeaders = (): EntityHeader[] => {
+const getStoredEntityHeaders = (): EntityHeader[] => {
```
Comment (nickgros) [inline src/lib/utils/functions/getDataFromFromStorage.ts:9]: Don't want to encourage using these, so don't export anything that we don't need to export.

File: src/lib/utils/functions/getDataFromFromStorage.ts [hunk 15]
```diff
@@ -33,3 +38,39 @@ export const getStoredEvaluation = (): Evaluation[] => {
...
+      return getDisplayValueForEntityColumn(value)
+    case ColumnType.USERID:
+    case ColumnType.USERID_LIST:
+      return getDisplayValueUserIdColumn(value)
+    case ColumnType.EVALUATIONID:
+      return getDisplayValueEvaluationIdColumn(value)
+    default:
+      return value
+  }
+}
```
Comment (nickgros) [inline src/lib/utils/functions/getDataFromFromStorage.ts:76]: Combined the VALUE_NOT_SET rule with fetching by ID from the sessionStorage cache

File: src/lib/containers/widgets/ElementWithTooltip.tsx [hunk 16]
```diff
@@ -28,7 +28,6 @@ export type TooltipVisualProps = {
 type ElementWithTooltipProps = React.PropsWithChildren<{
   image?: IconSvgOptions | CustomImageProps
   imageColor?: string
-  idForToolTip: string
```
Comment (nickgros) [inline src/lib/containers/widgets/ElementWithTooltip.tsx:31]: I don't think we need this prop anymore since we switched to the MUI tooltip

**Review by jay-hodgson — APPROVED** [review]

**Pull request merged**